grails-app/services/org/pih/warehouse/data/LocationDataService.groovy

```diff
@@ -33,32 +31,44 @@ class LocationDataService {
 
         command.data.eachWithIndex { params, index ->
 
-            Location location = createOrUpdateLocation(params)
-
+            Location newLocation = Location.findByName(params.name)
+            Location location = new Location()
+            Location parentLocation = params.parentLocation ? Location.findByName(params.parentLocation) : null
             LocationGroup locationGroup = params.locationGroup ? LocationGroup.findByName(params.locationGroup) : null
+            LocationType locationType = params.locationType ? LocationType.findByName(params.locationType) : null
+
+            if (newLocation) {
+                location.errors.reject("Row ${index + 1}: '${params.name}' already exists in the system. Choose a unique name")
+            }
+
+            if (params.parentLocation && !parentLocation) {
+                location.errors.reject("Row ${index + 1}: Parent Location '${params.parentLocation}' does not exist")
+            }
+
             if (params.locationGroup && !locationGroup) {
-                command.errors.reject("Row ${index + 1}: Location group '${params.locationGroup}' does not exist")
+                location.errors.reject("Row ${index + 1}: Location Group '${params.locationGroup}' does not exist")
+            }
+
+            if (!locationType) {
+                location.errors.reject("Row ${index + 1}: Location Type '${params.locationType}' does not exist")
             }
 
-            if (!location.validate()) {
-                location.errors.each { BeanPropertyBindingResult error ->
-                    String errorMessage = messageSource.getMessage(error.fieldError, locale)
-                    command.errors.reject("Row ${index + 1}: ${errorMessage}")
-                }
+            if (locationType?.locationTypeCode in [LocationTypeCode.INTERNAL, LocationTypeCode.BIN_LOCATION, LocationTypeCode.ZONE] && !parentLocation) {
```

Comment:
Let's move the list of LocationTypeCodes into a static method on LocationType
```
LocationType.getInternalLocationTypeCodes() { return [LocationTypeCode.INTERNAL, LocationTypeCode.BIN_LOCATION, LocationTypeCode.ZONE] }
LocationType.getFacilityLocationTypeCodes() { ... }
```
And/or something more DRY like a method a boolean method that returns whether the location is an internal location or facility level location
```
Boolean isInternalLocation() { return locationTypeCode in LocationType.getInternalLocationTypeCodes() }
Boolean isFacilityLocation() { return locationTypeCode in LocationType.getFacilityLocationTypeCodes() } 
```
to prevent us from having to copy that array all over the place.

Reply:
Sure, gonna correct it.